app/src/main/java/com/azuresamples/msalandroidapp/MainActivity.java

```diff
@@ -0,0 +1,552 @@
+        TextView apiText = (TextView) findViewById(R.id.apiData);
+        apiText.setText("API Response: " + apiResponse.toString());
+    }
+
+    /* Set the UI for successful token acquisition data */
+    private void updateSuccessUI() {
+        // hide the old buttons
+        loginButton.setVisibility(View.INVISIBLE);
+
+        // display sign out and edit buttons
+        signOutButton.setVisibility(View.VISIBLE);
+        editProfileButton.setVisibility(View.VISIBLE);
+        findViewById(R.id.welcome).setVisibility(View.VISIBLE);
+        ((TextView) findViewById(R.id.welcome)).setText(
+                "Welcome!\n\n\n\nUser's unique identifier\n\n" +
+                authResult.getAccount().getAccountIdentifier().getIdentifier());
+        findViewById(R.id.apiData).setVisibility(View.VISIBLE);
+    }
+
+    /* Set the UI for signed out account */
+    private void updateSignedOutUI() {
+        loginButton.setVisibility(View.VISIBLE);
+
+        editProfileButton.setVisibility(View.INVISIBLE);
+        signOutButton.setVisibility(View.INVISIBLE);
+        findViewById(R.id.welcome).setVisibility(View.INVISIBLE);
+        findViewById(R.id.apiData).setVisibility(View.INVISIBLE);
+        ((TextView) findViewById(R.id.apiData)).setText("No Data");
+    }
+
+    //
+    // App callbacks for MSAL
+    // ======================
+    // getActivity() - returns activity so we can acquireToken within a callback
+    // getAuthSilentCallback() - callback defined to handle acquireTokenSilent() case
+    // getAuthInteractiveCallback() - callback defined to handle acquireToken() case
+    // getResetCallback() - callback defined to handle the acquireToken() for reset password
+    // getEditCallback() - callback defined to handle the acquireToken() for edit profile
+
+    public Activity getActivity() {
+        return this;
+    }
+
+    /* Callback used in for silent acquireToken calls.
+     * Looks if tokens are in the cache (refreshes if necessary and if we don't forceRefresh)
+     * else errors that we need to do an interactive request.
+     */
+    private AuthenticationCallback getAuthSilentCallback() {
+        return new AuthenticationCallback() {
+            @Override
+            public void onSuccess(AuthenticationResult authenticationResult) {
+                /* Successfully got a token, call API now */
+                Log.d(TAG, "Successfully authenticated");
+
+                /* Store the authResult */
+                authResult = authenticationResult;
+
+                /* call API */
+                callApi();
+
+                /* update the UI to post call API state */
+                updateSuccessUI();
+            }
+
+            @Override
+            public void onError(MsalException exception) {
+                /* Failed to acquireToken */
+                Log.d(TAG, "Authentication failed: " + exception.toString());
+
+                if (exception instanceof MsalClientException) {
```

Comment:
Logging the exception error message would be helpful i believe.

Also, not necessarily in this PR but in future may be we want to call `acquireToken` when `MsalUiRequiredException` is received